clippy_lints/src/assigning_clones.rs

```diff
@@ -0,0 +1,204 @@
+use clippy_utils::{diagnostics::span_lint_and_then, sugg::Sugg, ty::implements_trait};
+use rustc_errors::Applicability;
+use rustc_hir::{self as hir, ExprKind};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+use rustc_span::symbol::sym;
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for code like `foo = bar.clone();`
+    ///
+    /// ### Why is this bad?
+    /// If cloning `bar` allocates memory (or other resources), then this code will allocate
+    /// new memory for the clone of `bar`, then drop `foo`, then overwrite `foo` with the clone.
+    /// Instead, `Clone::clone_from()`, or `ToOwned::clone_into()`, may be able to update
+    /// `foo` in-place, reusing existing memory.
+    ///
+    /// Note that this only provides any actual improvement if the type has explicitly implemented
+    /// the `clone_from()` trait method, since the trait-provided implementation will just call
+    /// `clone()`.
+    ///
+    /// ### Example
+    /// ```rust
+    /// #[derive(Clone)]
+    /// struct Thing;
+    ///
+    /// pub fn assign_to_ref(a: &mut Thing, b: Thing) {
+    ///     *a = b.clone();
+    /// }
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// #[derive(Clone)]
+    /// struct Thing;
+    ///
+    /// pub fn assign_to_ref(a: &mut Thing, b: Thing) {
+    ///     a.clone_from(&b);
+    /// }
+    /// ```
```

Comment:
```suggestion
    /// ### Example
    /// ```rust
    /// # let src = "Cheesecake".to_string();
    /// let mut target = "Hello world".to_string();
    /// // [...]
    /// target = src.clone();
    /// ```
    /// Use instead:
    /// ```rust
    /// # let src = "Cheesecake".to_string();
    /// let mut target = "Hello world".to_string();
    /// // [...]
    /// target.clone_from(&src);
    /// ```
```